www/caddy/src/opnsense/scripts/OPNsense/Caddy/caddy_certs.php

```diff
@@ -27,61 +27,83 @@
  *    POSSIBILITY OF SUCH DAMAGE.
  */
 
+namespace OPNsense\Caddy;
+
 require_once "config.inc";
 
-use OPNsense\Core\Config;
-
-$configObj = Config::getInstance()->object();
-$temp_dir = '/var/db/caddy/data/caddy/certificates/temp/';
-
-// Traverse through certificates
-foreach ($configObj->cert as $cert) {
-    $cert_refid = (string)$cert->refid;
-    $cert_content = base64_decode((string)$cert->crt);
-    $key_content = base64_decode((string)$cert->prv);
-    $cert_chain = $cert_content;
-
-    // Handle CA and possible intermediate CA to create a certificate bundle
-    if (!empty($cert->caref)) {
-        foreach ($configObj->ca as $ca) {
-            if ((string)$cert->caref === (string)$ca->refid) {
-                $ca_content = base64_decode((string)$ca->crt);
-                $cert_chain .= "\n" . $ca_content;
-
-                if (!empty($ca->caref)) {
-                    foreach ($configObj->ca as $parent_ca) {
-                        if ((string)$ca->caref === (string)$parent_ca->refid) {
-                            $parent_ca_content = base64_decode((string)$parent_ca->crt);
-                            $cert_chain .= "\n" . $parent_ca_content;
-                            break;
-                        }
-                    }
-                }
+use OPNsense\Trust\Ca;
+use OPNsense\Trust\Cert;
+use OPNsense\Trust\Store as CertStore;
+
+$writeFileIfChanged = function ($filePath, $content) {
+    if (
+        !file_exists($filePath) ||
+        hash('sha256', $content) !== hash_file('sha256', $filePath)
+    ) {
+        file_put_contents($filePath, $content);
+    }
+};
+
+$tempDir = '/var/db/caddy/data/caddy/certificates/temp/';
+
+// leaf certificate chain
+$certificateRefs = [];
+
+foreach ((new Caddy())->reverseproxy->reverse->iterateItems() as $reverseItem) {
+    $certRef = (string)$reverseItem->CustomCertificate;
+    if (!empty($certRef)) {
+        $certificateRefs[] = $certRef;
+    }
+}
+
+$certificateRefs = array_unique($certificateRefs);
+
+foreach ((new Cert())->cert->iterateItems() as $cert) {
+    $refid = (string)$cert->refid;
+
+    if (in_array($refid, $certificateRefs, true)) {
+        $certChain = base64_decode((string)$cert->crt);
+        $certKey = base64_decode((string)$cert->prv);
+
+        if (!empty((string)$cert->caref)) {
+            $ca = CertStore::getCACertificate((string)$cert->caref);
```

Comment:
I think you want to collect the full chain here, in which case `CertStore::getCaChain` would be more appropriate.

Reply:
Wow thanks I missed that. Its exactly what I needed. :)